8253952: Refine ZipOutputStream.putNextEntry() to recalculate ZipEntry's compressed size

simonis opened the pull request:
### Summary

Work around wrong usage of `ZipOutputStream.putNextEntry()` in user code which can lead to the `ZipException "invalid entry compressed size"`.

### Motivation

In general it is not safe to directly write a ZipEntry obtained from `ZipInputStream.getNextEntry()`, `ZipFile.entries()`, `ZipFile.getEntry()` or `ZipFile.stream()` with `ZipOutputStream.putNextEntry()` to a `ZipOutputStream` and then read the entries data from the `ZipInputStream` and write it to the `ZipOutputStream` as follows:
```
 ZipEntry entry;
 ZipInputStream zis = new ZipInputStream(...);
 ZipOutputStream zos = new ZipOutputStream(...);
 while((entry = zis.getNextEntry()) != null) {
     zos.putNextEntry(entry);
     zis.transferTo(zos);
 }
```
The problem with this code is that the zip file format does not record the compression level used for deflation in its entries. In general, it doesn't even mandate a predefined compression ratio per compression level. Therefore the compressed size recorded in a `ZipEntry` read from a zip file might differ from the new compressed size produced by the receiving `ZipOutputStream`. Such a difference will result in a `ZipException` with the following message:
```
 java.util.zip.ZipException: invalid entry compressed size (expected 12 but got 7 bytes)
``` 
The correct way of copying all entries from one zip file into another requires the creation of a new `ZipEntry` or at least resetting of the compressed size field. E.g.:
```
 while((entry = zis.getNextEntry()) != null) {
     ZipEntry newEntry = new ZipEntry(entry.getName());
     zos.putNextEntry(newEntry);
     zis.transferTo(zos);
 }
```
or:
```
 while((entry = zis.getNextEntry()) != null) {
     entry.setCompressedSize(-1);
     zos.putNextEntry(entry);
     zis.transferTo(zos);
 }
```
Unfortunately, there's a lot of user code out there which gets this wrong and uses the bad coding pattern described before. Searching for `"java.util.zip.ZipException: invalid entry compressed size (expected 12 but got 7 bytes)"` gives ~2500 hits (~100 on StackOverflow). It's also no hard to find plenty of instances of this anti-pattern on GitHub when doing a code search for `ZipEntry` and `putNextEntry()`. E.g. [Gradle 4.x wrapper task][1] is affected as well as the latest version of the [mockableAndroidJar task][2]. I've recently fixed two occurrences of this pattern in OpenJDK (see [JDK-8240333][3] and [JDK-8240235][4]) but there still exist more of them (e.g. [`test/jdk/java/util/zip/ZipFile/CopyJar.java`][5] which is there since 1999 :).

### Description

So while this has clearly been a problem before, it apparently wasn't painful enough to trigger any action from the side of the JDK. However, recently quite some zlib forks with [superior deflate/inflate performance have evolved][6]. Using them with OpenJDK is quite straight-forward: one just has to configure the alternative implementations by setting `LD_LIBRARY_PATH` or `LD_PRELOAD` correspondingly. We've seen big saving by using these new zlib implementations for selected services in production and the only reason why we haven't enabled them by default until now is the problem I've just described. The reason why these new libraries uncover the described anti-pattern much more often is because their compression ratio is slightly different from that of the default zlib library. This can easily trigger a `ZipException` even if an application is not using a different compression levels but just a zip file created with another zlib version.

I'd therefore like to propose the following workaround for the wrong `ZipOutputStream.putNextEntry()` usage in user code:

-  ignore the compressed size if it was implicitly determined from the zip file and not explicitly set by calling `ZipEntry.setCompressedSize()`.

- Change the API-documentation of `ZipOutputStream.putNextEntry()` and `JarOutputStream.putNextEntry()` to explain the problem and why `putNextEntry()` will ignore the compressed size of a `ZipEntry` if that was set implicitely when reading that entry from a `ZipFile` or `ZipInputStream`.


### Technical Details

A zip file consists of a stream of File Entries followed by a Central Directory (see [here for a more detailed specification][7]). Each File Entry is composed of a Local File Header (LFH) followed by the compressed Data and an optional Data Descriptor. The LFH contains the File Name and among other attributes the Compressed and Uncompressed size and CRC of the Data. In the case where the latter three attributes are not available at the time when the LFH is created, this fact will be recorded in a flag of the LFH and will trigger the creation of a Data Descriptor with the corresponding information right after the Data section. Finally, the Central Directory contains one Central Directory File Header (CDFH) for each entry of the  zip archive. The CDFH is an extended version of the LFH and the ultimate reference for the contents of the zip archive. The redundancy between LFH and CDFH is a tribute to zip's long history when it was used to store archives on multiple floppy discs and the CDFH allowed to update the archive by only writing to the last disc which contained the Central Directory.

`ZipEntries` read with `ZipInputStream.getNextEntry()` will initially only contain the information from the LFH. Only after the next entry was read (or after `ZipInputStream.closeEntry()` was called explicitly), will the previously read entry be updated with the data from the Data Descriptor. `ZipInputStream` doesn't inspect the Central Directory at all.

On the other hand, `ZipFile` only queries the Central Directory for `ZipEntry` information so all `ZipEntries` returned by `ZipFile.entries()`, `ZipFile.getEntry()` and `ZipFile.stream()` will always instantly contain the full Compressed and Uncompressed Size and CRC information for each entry independently of the LFH contents.

### Risks and Assumptions

If we choose to ignore the implicitly recorded compressed size in a `ZipEntry` read from a zip file when writing it to a `ZipOutputStream`, this will lead to zip files with incomplete information in the LFH and an additional Data Descriptor as described before. However, the result is still fully compatible to the zip file specification. It's also not unusual, because by default all new zip files created with `ZipOutputStream` will contain LFHs without Compressed and Uncompressed Size and CRC information and an additional Data Descriptor. Theoretically it is possible to create new zip files with `ZipOutputStream` class and Compressed and Uncompressed Size and CRC information in the LFH but that's complex and inefficient because it requires two steps. A first step to determine the crc and compressed size of the data and a second step to actually write the data to the `ZipOutputStream` (which will compress it a second time). This is because the current API offers no possibility to write already compressed data to a `ZipOutputStream`.

Consequently, the only straight-forward way of creating zip files from Java which have all the data in the LFH and no Data Descriptor is by copying `ZipEntries` from an existing zip file with the buggy method described before. This incidentally worked more or less reliable for a long time but breaks miserably when using different zlib implementations. Ignoring the implicitly set compressed size of `ZipEntries` can easily fix this problem.

I'm not aware of any tool which can not handle such files and if it exists it would have problems with the majority of Java created zip files anyway (e.g. all jar-files created with the jar tool have zip entries with incomplete LFH data and additional Data Descriptor).

Ignoring the implicitly set compressed size of `ZipEntries` has no measurable performance impact and will increase the size of zip archives which used to have the complete file information in the LFH before by 16 bytes per entry. On the other hand it will give us the freedom to use whatever zip implementation we like :)


[1]: https://github.com/gradle/gradle/blob/e76905e3a/subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java#L152-L158
[2]: https://android.googlesource.com/platform/tools/base/+/refs/heads/master/build-system/builder/src/main/java/com/android/builder/testing/MockableJarGenerator.java#86
[3]: https://bugs.openjdk.java.net/browse/JDK-8240333
[4]: https://bugs.openjdk.java.net/browse/JDK-8240235
[5]: https://github.com/openjdk/jdk/blob/master/test/jdk/java/util/zip/ZipFile/CopyJar.java
[6]: https://github.com/simonis/zlib-bench/blob/master/Results.md
[7]: https://en.wikipedia.org/wiki/Zip_(file_format)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8253952](https://bugs.openjdk.java.net/browse/JDK-8253952): Refine ZipOutputStream.putNextEntry() to recalculate ZipEntry's compressed size


### Reviewers
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 7f032e0ff5064a3f29aab8434d96362dccda6626


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/520/head:pull/520`
`$ git checkout pull/520`

LanceAndersen commented on src/java.base/share/classes/java/util/zip/ZipOutputStream.java:
```diff
@@ -182,6 +182,21 @@ public void setLevel(int level) {
      * was specified for the entry, and the current time will be used if
      * the entry has no set modification time.
+     *
+     * The zip file format does not record the compression level used for the
```
"zip" ->"ZIP" for consistency'

Should this refer to the Zip File Format specification?

AlanBateman commented on src/java.base/share/classes/java/util/jar/JarOutputStream.java:
```diff
@@ -81,6 +81,20 @@ public JarOutputStream(OutputStream out) throws IOException {
+     * this output stream will ignore the compressed size of a {@code ZipEntry}
+     * and re-compute its value automatically after the associted data has been
+     * completely deflated.
+     *
```
We probably need to do a bit wordsmithing here.

I think I'd drop the first two sentences and instead start a new paragraph here (<p> tag) with "Unless explicitly set, the output stream will ignore ..." and see how that looks.

Probably should be "the ZipEntry" rather than "a ZipEntry" to be consistent with the parameter description.

LanceAndersen replied:
Alan makes a good point.  Perhaps we keep things simple and focus solely on tweaking the description of the change in behavior which is described in your last paragraph

simonis replied:
I totally agree. What about using just the last sentence (as you've proposed) in the spec section and add the other to as @implNote? O you think the last sentence will be enough?

LanceAndersen replied:
I think we can just go with the last sentence/paragraph.  Perhaps  we can further simplify the  paragraph/sentence with something like:

The compressed entry size will be recalculated  for compressed (DEFLATED) entries when  ZipEntry::setCompressedSize has not been explicitly called on the ZipEntry.

or 

The compressed (DEFLATED) entry size will be recalculated when  ZipEntry::setCompressedSize has not been explicitly called on the ZipEntry.

LanceAndersen replied:
I think the wording looks much better now

AlanBateman commented:
/csr

LanceAndersen commented:
I think as we start to move forward with the review and CSR, we should update the bug and PR description as the change is not a workaround but a change in behavior of the implementation

simonis commented:
I already changed the bug description on Alan's request and just updated the PR description to reflect that. Please feel free to propose a better description.

simonis commented:
Thanks for your feedback. I've updated the API doc as requested and created a CSR at:

https://bugs.openjdk.java.net/browse/JDK-8254284

Can somebody please also review the CSR?

Thank you and best regards,
Volker

LanceAndersen commented on src/java.base/share/classes/java/util/jar/JarOutputStream.java:
```diff
@@ -81,6 +81,10 @@ public JarOutputStream(OutputStream out) throws IOException {
      * The current time will be used if the entry has no set modification
      * time.
      *
+     * The compressed size field will be recalculated for compressed (i.e.
```
I think the wording is better, but not sure we should state "compressed size field" as ZipEntry.setCompressedSize refers to "Sets the size of the compressed entry data."   and "the compressed size to set" 

I think I would omit "field" from the above so it reads  similar to "The compressed size will be..." or "The compressed size value will be..."

AlanBateman commented on src/java.base/share/classes/java/util/zip/ZipOutputStream.java:
```diff
@@ -182,6 +182,11 @@ public void setLevel(int level) {
+     * The compressed size field will be recalculated for compressed (i.e.
+     * {@code ZipEntry.DEFLATED}) entries when {@link ZipEntry#setCompressedSize(long)}
+     * has not been explicitly called on the {@code ZipEntry}.
+     *
```
Here's an alternative that might be a bit clearer to readers.

"When writing a compressed (deflated) entry, and the compressed size has not been explicitly set with the setCompressedSize method, then the compressed size written to the entry's data descriptor will be its actual compressed size."

I think we should put it after the "The default compression method .." sentence and move the sentence on the time stamp to the end.

LanceAndersen replied:
I don't believe we discuss/reference the data descriptor for a Zip entry (outside of the PKWare Zip specification) so I am not sure we should reference it in the javadoc. 

 Placing the sentence after "The default compression method will be used if no compression method was specified for the entry" makes sense

simonis replied:
Thanks for your input. I've tried to somehow merge both suggestions :)

Did you had a chance to look at the CSR? I think somebody has to review it before I can move it to "Finalized".

Thank you and best regards,
Volker

LanceAndersen replied:
I have added myself as a reviewer based on the updates made by you and Alan

AlanBateman commented on src/java.base/share/classes/java/util/jar/JarOutputStream.java:
```diff
@@ -78,8 +78,12 @@ public JarOutputStream(OutputStream out) throws IOException {
+     * the {@link ZipEntry#setCompressedSize(long)} method, then the
+     * compressed size will be set to the actual compressed size after
+     * deflation. The current time will be used if the entry has no set
+     * modification time.
```
Thanks for combining the wording, I think this looks good. I'd probably drop "i.e." so that it's just "DEFLATED" parentheses.

LanceAndersen replied:
I might consider making "The current time..." its own paragraph separate from the compression discussion as I think it would be clearer that way.

AlanBateman commented on src/java.base/share/classes/java/util/zip/ZipEntry.java:
```diff
@@ -53,6 +53,7 @@
     long crc = -1;      // crc-32 of entry data
     long size = -1;     // uncompressed size of entry data
     long csize = -1;    // compressed size of entry data
+    boolean manual_csize = false; // Only true if csize was explicitely set by a call to setCompressedSize()
```
Minor nit but this should probably be "csizeSet" as it's not too common to have underscore in field names.

AlanBateman commented on src/java.base/share/classes/java/util/zip/ZipOutputStream.java:
```diff
@@ -205,9 +210,15 @@ public void putNextEntry(ZipEntry e) throws IOException {
                 e.flag = 8;
-
+            }
+            else if (!e.manual_csize) {
```
I assume the existing expression that set if csize has been set so that we don't set the flag to 8 in two branches.

LanceAndersen replied:
I was thinking about that myself

AlanBateman commented on test/jdk/java/util/zip/CopyZipFile.java:
```diff
@@ -0,0 +1,204 @@
+    private static void createZip(String zipFile) throws Exception {
+        File f = new File(zipFile);
+        f.deleteOnExit();
+        OutputStream os = new FileOutputStream(f);
```
Can you use try-with-resources here so that a test failure will close the file.

LanceAndersen commented on src/java.base/share/classes/java/util/jar/JarOutputStream.java:
```diff
@@ -78,8 +78,12 @@ public JarOutputStream(OutputStream out) throws IOException {
      * used if no compression method was specified for the entry.
-     * The current time will be used if the entry has no set modification
-     * time.
+     * When writing a compressed (i.e. {@code ZipEntry.DEFLATED}))
```
I would leave as DEFLATED given it is used this way in other ZipOutputStream methods and I would also remove the  "i.e."

AlanBateman replied:
I would prefer to keep it as "When writing as compressed (DEFLATED) ..." to keep the terminology consistent when linking to the setCompressedMethod.

LanceAndersen replied:
Agree, no need for ZipEntry.DEFLATED which is what I was trying to say (sorry if that was not clear).  Only suggesting ZipEntry.DEFLATED -> DEFLATED and remove  the ".i.e."

LanceAndersen reviewed (changes requested):
I think we are getting close.    Once the javadoc is cleaned up along with the minor changes recommended  we should be in good shape and can then review the CSR

simonis commented:
I hope I've addressed all your comments and suggestion with the latest PR. I've also updated the CSR to reflect the latest version of the API changes from the PR. Please have a look.

Thank you and best regards,
Volker

LanceAndersen commented on test/jdk/java/util/zip/CopyZipFile.java:
```diff
@@ -0,0 +1,204 @@
+        // or ZipInputStream.
+        os = new ByteArrayOutputStream();
+        zos = new ZipOutputStream(os);
+        ZipFile zf = new ZipFile(ZIP_FILE);
```
Any reason to not use try with resources here as well (and below)

simonis replied:
Done

AlanBateman commented on src/java.base/share/classes/java/util/jar/JarOutputStream.java:
```diff
@@ -78,6 +78,11 @@ public JarOutputStream(OutputStream out) throws IOException {
+     * size has not been explicitly set with the {@link
+     * ZipEntry#setCompressedSize(long)} method, then the compressed
+     * size will be set to the actual compressed size after deflation.
+     * <p>
```
I'm happy with the wording. What you think about put a p tag before "The default compression method ..." so that it goes into its own paragraph? I'm asking because it's inconsistent to have the first paragraph include the details on the compression method and the details on the current time pushed into a second paragraph - if you generate the javadoc then you'll see what I mean.

LanceAndersen replied:
I think that is a good idea to at the paragraph tag as you suggest

simonis replied:
Done

AlanBateman commented on src/java.base/share/classes/java/util/zip/ZipEntry.java:
```diff
@@ -447,6 +450,7 @@ public long getCompressedSize() {
      */
     public void setCompressedSize(long csize) {
         this.csize = csize;
+        this.csizeSet = true;
```
Latest version of implementation changes look good.

LanceAndersen commented on test/jdk/java/util/zip/CopyZipFile.java:
```diff
@@ -0,0 +1,203 @@
+        // therefore the ZipEntry objects created by ZipOutputStream.getNextEntry() will have
+        // all these fields set to '-1'.
+        InputStream is = new FileInputStream(ZIP_FILE);
+        ZipInputStream zis = new ZipInputStream(is);
```
Any reason not to include the the ZipInputStream and InputStream in the try with Resources and the ZipFile below?  I know these are nits, but just curious or was it an over site?

I think we are good otherwise

simonis replied:
No other reason except my laziness :)

simonis replied:
fixed now

LanceAndersen commented on test/jdk/java/util/zip/CopyZipFile.java:
```diff
@@ -0,0 +1,203 @@
+        // Directory which must have accurate information for size, compressedSize and crc.
+        // This means that all ZipEntry objects returned from ZipFile will have correct
+        // settings for these fields.
+        // I the compression level was different in the initial zip file (which we can't find
```
typo I -> If

simonis replied:
Fixed

LanceAndersen commented on test/jdk/java/util/zip/CopyZipFile.java:
```diff
@@ -0,0 +1,203 @@
+                } catch (ZipException ze) {
+                    if ("test1.txt".equals(entry.getName()) || "test2.txt".equals(entry.getName())) {
+                        throw new Exception(
+                            "Should throw for STORED files or files compressed with DEFAULT_COMPRESSION");
```
Perhaps clean up the message above "Should throw" to indicate what is being thrown

simonis replied:
Thanks. That should actually red "Shouldn't...". Fixed that and also added ZipExcpetion as exception which shouldn't be thrown.

LanceAndersen reviewed (approved):
The changes look good.  Thank you for the updates.  I have kicked off mach5 jdk-tier1,jdk-tier2,jdk-tier3 runs across all of the platforms.

simonis commented:
> Mach5 run is clean :-)

Thanks a lot Lance.
Just waiting for the CSR to get approved now.

simonis commented:
/integrate
